File: shopping_mall/backend/app/models/product.py
from __future__ import annotations
from typing import TYPE_CHECKING, List, Optional
from datetime import datetime
from sqlalchemy import DateTime, Float, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database import Base

if TYPE_CHECKING:
    from .category import Category
    from .store import Store
    from .review import Review


class Product(Base):
    __tablename__ = "products"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(String(300), nullable=False)
    description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    price: Mapped[int] = mapped_column(Integer, nullable=False)
    discount_rate: Mapped[int] = mapped_column(Integer, default=0)
    category_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("categories.id"), nullable=True)
    store_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("stores.id"), nullable=True)
    thumbnail: Mapped[Optional[str]] = mapped_column(String(500), nullable=True)
    images: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    options: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    stock: Mapped[int] = mapped_column(Integer, default=0)
    rating: Mapped[float] = mapped_column(Float, default=0.0)
    review_count: Mapped[int] = mapped_column(Integer, default=0)
    sales_count: Mapped[int] = mapped_column(Integer, default=0)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())

    category: Mapped[Optional["Category"]] = relationship("Category", back_populates="products")
    store: Mapped[Optional["Store"]] = relationship("Store", back_populates="products")
    reviews: Mapped[List["Review"]] = relationship("Review", back_populates="product")

File: shopping_mall/frontend/src/types/product.ts
import type { Category } from './category';
import type { Store } from './store';

export interface Product {
  id: number;
  name: string;
  description: string | null;
  price: number;
  discountRate: number;
  thumbnail: string | null;
  images: string[];
  options: ProductOption[];
  stock: number;
  rating: number;
  reviewCount: number;
  salesCount: number;
  createdAt: string;
  category?: Category;
  store?: Store;
  storeName?: string;
}

export interface ProductOption {
  name: string;
  values: string[];
}

export interface ProductListResponse {
  items: Product[];
  total: number;
  page: number;
  limit: number;
  totalPages: number;
}
